daemon/delete.go

```diff
@@ -141,7 +141,10 @@ func (daemon *Daemon) cleanupContainer(container *container.Container, config ba
 			container.SetRemovalError(err)
 			return err
 		}
+		// Hold the container lock while clearing RWLayer reference to avoid dereferencing nil in other places.
+		container.Lock()
 		container.RWLayer = nil
+		container.Unlock()
```

Comment:
It's probably a bit inefficient as right after this if-else statement a lock is acquired again. But I'm not sure if it's worth to avoid an extra locking as the code would get more complicated. WDYT?

Reply:
I think we could just add 
```go
rwLayer := container.RWLayer
contianer.RWLayer = nil
```
before L134 `container.Unlock()` and then change the the remaining `container.RWLayer` references to the `rwLayer` local variable.

Reply:
That would change the behaviour when `daemon.imageService.ReleaseLayer` returns an error - the RWLayer would be cleared from the container but would not be released. Is that ok?

Reply:
Hmm right, in this case I think we can just restore it (with extra lock) when returning the error. This way we need to acquire the lock again only in case the release fail (it really _shouldn't_).

Reply:
I updated the PR accordingly